File: server/models/game.py
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from models.user import UserOut


class GameBase(BaseModel):
    opponent: str
    dictionary: str = "TWL06"
    board_type: str = "standard"
    turn_timer: bool = False
    duration: Optional[int] = None
    timeIncrement: Optional[int] = None
    online: bool = True
    disputes: bool = False
    dispute_timeout: int = 60


class TilePlacement(BaseModel):
    row: int
    col: int
    letter: str


class GameMove(BaseModel):
    move_number: int
    move_type: str
    player: Optional[str] = None
    tiles: List[TilePlacement] = []
    recycled: List[str] = []
    score: int = 0
    word: str = ""
    all_words: List[str] = []
    rack: List[str] = []
    tile_bag: List[str] = []
    timestamp: str
    dispute_status: Optional[str] = None


class PendingDispute(BaseModel):
    player: str
    opponent: str
    score: int
    word: str
    all_words: List[str] = []
    expires_at: str
    tiles: List[dict] = []
    rack_before: List[str] = []
    move_number: int = 0
    timestamp: str = ""


class GameState(BaseModel):
    game_id: str
    players: List[str]
    board: List[List[Optional[str]]]
    tile_bag: List[str]
    racks: Dict[str, List[str]]
    scores: Dict[str, int]
    turn: str
    status: str
    winner: Optional[str] = None
    game_moves: List[GameMove] = []
    pending_dispute: Optional[PendingDispute] = None


class GameStateSummary(BaseModel):
    turn: Optional[str] = None
    status: str = "waiting"
    last_move_at: Optional[str] = None
    scores: Dict[str, int] = {}


class GameOut(BaseModel):
    id: str
    user: UserOut
    opponent: UserOut
    dictionary: str = "TWL06"
    board_type: str = "standard"
    turn_timer: bool = False
    duration: Optional[int] = None
    timeIncrement: Optional[int] = None
    online: bool = True
    disputes: bool = False
    completed: bool = False
    winner: Optional[str] = None
    loser: Optional[str] = None
    userScore: int = 0
    opponentScore: int = 0
    date: Optional[datetime] = None
    game_state: Optional[GameStateSummary] = None

File: app/app/services/game.ts
import { apiFetch } from './api';
import type { GameUser } from './user';

export type { GameUser } from './user';

export interface CreateGameParams {
  opponent: string;
  dictionary: string;
  board_type: string;
  turn_timer: boolean;
  duration: number | null;
  timeIncrement: number | null;
  online: boolean;
  disputes: boolean;
  dispute_timeout: number;
}

export interface GameStateSummary {
  turn: string | null;
  status: 'waiting' | 'active' | 'finished';
  last_move_at: string | null;
  scores: Record<string, number>;
}

export interface Game {
  id: string;
  user: GameUser;
  opponent: GameUser;
  dictionary: string;
  board_type: string;
  turn_timer: boolean;
  duration: number | null;
  timeIncrement: number | null;
  online: boolean;
  disputes: boolean;
  dispute_timeout: number;
  completed: boolean;
  winner: string | null;
  loser: string | null;
  userScore: number;
  opponentScore: number;
  date: string;
  game_state: GameStateSummary | null;
}

export interface PendingDispute {
  player: string;
  opponent: string;
  score: number;
  word: string;
  all_words: string[];
  expires_at: string; // ISO datetime string
}

export interface GameMove {
  move_number: number;
  move_type: 'initial' | 'place' | 'pass' | 'recycle' | 'resign' | 'dispute';
  player: string | null;
  tiles: { row: number; col: number; letter: string; is_blank?: boolean }[];
  recycled: string[];
  score: number;
  word: string;
  all_words: string[];
  rack: string[];
  tile_bag: string[];
  timestamp: string;
  dispute_status?: 'not_disputed' | 'valid' | 'invalid';
}

export interface GameState {
  game_id: string;
  players: string[];
  board: (string | null)[][];
  tile_bag: string[];
  racks: Record<string, string[]>;
  scores: Record<string, number>;
  turn: string;
  status: 'waiting' | 'active' | 'finished';
  winner: string | null;
  game_moves: GameMove[];
  pending_dispute: PendingDispute | null;
}

export async function createGame(params: CreateGameParams, token: string): Promise<{ room: string }> {
  const data = await apiFetch<Game>('/game/create', {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify(params),
  });
  return { room: data.id };
}

export function getMyGames(token: string, completed?: boolean): Promise<Game[]> {
  const qs = completed !== undefined ? `?completed=${completed}` : '';
  return apiFetch<Game[]>(`/game/mine${qs}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function getGame(gameId: string, token: string): Promise<Game> {
  return apiFetch<Game>(`/game/${gameId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function getGameState(gameId: string, token: string): Promise<GameState> {
  return apiFetch<GameState>(`/game/${gameId}/state`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function getGameMoves(gameId: string, token: string): Promise<GameMove[]> {
  return apiFetch<GameMove[]>(`/game/${gameId}/moves`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}
